pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -89,7 +91,7 @@ func (w *watcher) Run(ctx context.Context) error {
 	return nil
 }
 
-// run periodically compares the image stored in the given provider and one stored in git.
+// run periodically compares the image in the given provider and one in git repository.
```

Comment:
Not in this PR's scope, but I feel that we should start thinking about simplifying the concurrency model implemented in this ImageWatcher package.

Current approach:

- Spawn goroutines for each image provider
- Inside each goroutine
    - loop the repositories
        - lock and pull to update the local git.repo data
        - load the configuration at `.pipe` to find the target providers
        - if the provider was specified, retrieve the latest images and update them

Suggestion:

- Spawn goroutines for each repository
- Inside each goroutine
   - pull to update the local repo data
   - load the configuration at `.pipe` to find the target providers
   - loop the target providers
       - retrieve the latest images and update them

Pros:
- don't need the look for `git.Pull`
- avoid complex concurrent processing for `git.Repo`: pull and push
- reduce the number of pulls and configuration loads (current: provider_num x repo_num, new: repo_num)

Cons:
- The pull interval for all image providers must be same (I suggest to move that field into `imageWatcher` part, `imageProvider` part is only for defining how to interact with the image provider)

Reply:
You're quite right. Current complex implementation is cause by the spec that requires to specify polling interval for each provider. I couldn't completely make up my mind for a long time. The way to be the same interval across all provider would be pretty simple, and love it. But isn't it really needed to configure the interval for each provider? I thought we should be more cautious about communication to the container registry than to git repository, that's why I designed like this.

Reply:
But as you said, using the same interval is pretty good if it's not worth supporting.

Reply:
For instance, a user has two image providers:
- Docker Hub that has the strict pull rate limit
- ECR that has not strict pull rate limit

For that, it's useful to respectively set pull intervals.

However, we don't have to handle such carefully if we define that case as an edge case.

Reply:
Yes, I think in that case they can split into multiple pipeds.

Reply:
Ah, agree with you. The general case had better be prioritized. Okay, let me work on them in another PR. I'll fix your other change requests.

Reply:
👍